File: app/src/tv/presets.ts
/**
 * Presets: sequências de botões gravadas por você, salvas por serviço.
 *
 * Existem porque não há API para legenda/áudio dentro do Max ou Disney+ — o menu é
 * desenhado pelo próprio app e só responde a botões. Em vez de chutar a sequência e
 * você calibrar no YAML, você grava navegando como sempre faz.
 *
 * Diferente do apps.yaml, isto aqui é escrito pela máquina: mora no armazenamento do
 * app como JSON, sem comentários para preservar.
 */

import { parseSteps, type Step } from "./macros.ts";

export const MAX_PRESETS_PER_SERVICE = 20;

export class PresetError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "PresetError";
  }
}

/** 'Legenda PT + áudio original' -> 'legenda-pt-audio-original'. */
export function slugify(label: string): string {
  const asciiOnly = label
    .normalize("NFKD")
    // Mesmo efeito do `encode("ascii", "ignore")` do Python: o que sobrou de acento
    // depois da decomposição sai fora, e "ç" já virou "c".
    .replace(/[^\x00-\x7F]/g, "")
    .toLowerCase();
  const slug = asciiOnly.replace(/[^a-z0-9]+/g, "-").replace(/^-+|-+$/g, "");
  return slug || "preset";
}

/** Uma sequência nomeada, pertencente a um serviço. */
export interface Preset {
  id: string;
  label: string;
  serviceId: string;
  steps: Step[];
  recordedAt?: string;
}

/** Serializa um passo no mesmo formato que o `apps.yaml` usa. */
export function stepToData(step: Step): Record<string, unknown> {
  switch (step.kind) {
    case "button": {
      const data: Record<string, unknown> = { button: step.name };
      if (step.times !== 1) data["times"] = step.times;
      if (step.delay !== 0.2) data["delay"] = step.delay;
      return data;
    }
    case "wait":
      return { wait: step.seconds };
    case "text":
      return { text: step.template };
    case "enter":
      return { enter: true };
    case "clear":
      return { clear: step.count };
  }
}

export function presetToData(preset: Preset): Record<string, unknown> {
  const data: Record<string, unknown> = {
    id: preset.id,
    label: preset.label,
    steps: preset.steps.map(stepToData),
  };
  if (preset.recordedAt) data["recorded_at"] = preset.recordedAt;
  return data;
}

/** Todos os presets, agrupados por serviço. */
export class PresetStore {
  constructor(readonly byService: Map<string, Preset[]> = new Map()) {}

  forService(serviceId: string): Preset[] {
    return [...(this.byService.get(serviceId) ?? [])];
  }

  get(serviceId: string, presetId: string): Preset | null {
    return this.forService(serviceId).find((p) => p.id === presetId) ?? null;
  }

  allPresets(): Preset[] {
    return [...this.byService.values()].flat();
  }

  /** Adiciona ou substitui. Regravar com o mesmo nome sobrescreve, que é o esperado. */
  put(preset: Preset): void {
    let presets = this.byService.get(preset.serviceId);
    if (!presets) {
      presets = [];
      this.byService.set(preset.serviceId, presets);
    }
    const index = presets.findIndex((existing) => existing.id === preset.id);
    if (index >= 0) {
      presets[index] = preset;
      return;
    }
    if (presets.length >= MAX_PRESETS_PER_SERVICE) {
      throw new PresetError(
        `Limite de ${MAX_PRESETS_PER_SERVICE} presets para ${preset.serviceId}. ` +
          "Apague algum antes de gravar outro.",
      );
    }
    presets.push(preset);
  }

  remove(serviceId: string, presetId: string): boolean {
    const presets = this.byService.get(serviceId);
    if (!presets || presets.length === 0) return false;
    const remaining = presets.filter((p) => p.id !== presetId);
    if (remaining.length === presets.length) return false;
    this.byService.set(serviceId, remaining);
    return true;
  }
}

export function parseStore(data: Record<string, unknown> | null | undefined): PresetStore {
  const byService = new Map<string, Preset[]>();
  for (const [serviceId, rawPresets] of Object.entries(data ?? {})) {
    if (!Array.isArray(rawPresets)) {
      throw new PresetError(`'${serviceId}' deveria ser uma lista de presets`);
    }
    const presets: Preset[] = rawPresets.map((entry) => {
      if (typeof entry !== "object" || entry === null || !("id" in entry)) {
        throw new PresetError(`preset sem 'id' em ${serviceId}: ${JSON.stringify(entry)}`);
      }
      const raw = entry as Record<string, unknown>;
      const preset: Preset = {
        id: String(raw["id"]),
        label: String(raw["label"] ?? raw["id"]),
        serviceId: String(serviceId),
        steps: parseSteps(raw["steps"] as Record<string, unknown>[] | undefined),
      };
      const recordedAt = raw["recorded_at"];
      if (typeof recordedAt === "string" && recordedAt) preset.recordedAt = recordedAt;
      return preset;
    });
    byService.set(String(serviceId), presets);
  }
  return new PresetStore(byService);
}

export function storeToData(store: PresetStore): Record<string, unknown> {
  const payload: Record<string, unknown> = {};
  for (const serviceId of [...store.byService.keys()].sort()) {
    const presets = store.byService.get(serviceId) ?? [];
    if (presets.length === 0) continue;
    payload[serviceId] = presets.map(presetToData);
  }
  return payload;
}

/** Carimbo em UTC, no formato que a UI mostra. */
export function nowStamp(at: Date = new Date()): string {
  return at.toISOString().slice(0, 16).replace("T", " ");
}

File: src/lgremote/tv/presets.py
"""Presets: sequências de botões gravadas por você, salvas por serviço.

Existem porque não há API para legenda/áudio dentro do Max ou Disney+ — o menu é
desenhado pelo próprio app e só responde a botões. Em vez de eu chutar a sequência e
você calibrar no YAML, você grava navegando como sempre faz.

Este arquivo é escrito pela máquina. O `apps.yaml` continua sendo o hand-written, com os
comentários que explicam como calibrar — uma regravação nunca encosta nele.
"""

from __future__ import annotations

import re
import unicodedata
from dataclasses import dataclass
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

import yaml

from lgremote.tv.macros import Step, parse_steps

MAX_PRESETS_PER_SERVICE = 20
_SLUG_STRIP = re.compile(r"[^a-z0-9]+")

_HEADER = """\
# Presets gravados pelo app (Gravar > navegar na TV > Parar).
#
# ATENÇÃO: este arquivo é REESCRITO pelo gravador — comentários que você
# adicionar aqui serão perdidos. Editar os passos à mão funciona, mas só
# até a próxima gravação daquele preset.
#
# Para calibrar a busca dentro dos apps, use o config/apps.yaml.
"""


class PresetError(ValueError):
    """Preset inválido ou arquivo malformado."""


def slugify(label: str) -> str:
    """'Legenda PT + áudio original' -> 'legenda-pt-audio-original'."""
    normalized = unicodedata.normalize("NFKD", label)
    ascii_only = normalized.encode("ascii", "ignore").decode("ascii").lower()
    slug = _SLUG_STRIP.sub("-", ascii_only).strip("-")
    return slug or "preset"


@dataclass(frozen=True)
class Preset:
    """Uma sequência nomeada, pertencente a um serviço."""

    id: str
    label: str
    service_id: str
    steps: tuple[Step, ...]
    recorded_at: str | None = None

    def as_yaml(self) -> dict[str, Any]:
        data: dict[str, Any] = {
            "id": self.id,
            "label": self.label,
            "steps": [step_to_yaml(step) for step in self.steps],
        }
        if self.recorded_at:
            data["recorded_at"] = self.recorded_at
        return data


def step_to_yaml(step: Step) -> dict[str, Any]:
    """Serializa um passo no mesmo formato que o `apps.yaml` usa."""
    from lgremote.tv.macros import ButtonStep, ClearStep, EnterStep, TextStep, WaitStep

    match step:
        case ButtonStep() as button:
            data: dict[str, Any] = {"button": button.name}
            if button.times != 1:
                data["times"] = button.times
            if button.delay != 0.2:
                data["delay"] = button.delay
            return data
        case WaitStep(seconds):
            return {"wait": seconds}
        case TextStep(template):
            return {"text": template}
        case EnterStep():
            return {"enter": True}
        case ClearStep(count):
            return {"clear": count}
    raise PresetError(f"Passo não serializável: {step!r}")


@dataclass
class PresetStore:
    """Todos os presets, agrupados por serviço."""

    by_service: dict[str, list[Preset]]

    def for_service(self, service_id: str) -> list[Preset]:
        return list(self.by_service.get(service_id, []))

    def get(self, service_id: str, preset_id: str) -> Preset | None:
        return next((p for p in self.for_service(service_id) if p.id == preset_id), None)

    def all_presets(self) -> list[Preset]:
        return [preset for presets in self.by_service.values() for preset in presets]

    def put(self, preset: Preset) -> None:
        """Adiciona ou substitui. Regravar com o mesmo nome sobrescreve, que é o esperado."""
        presets = self.by_service.setdefault(preset.service_id, [])
        for index, existing in enumerate(presets):
            if existing.id == preset.id:
                presets[index] = preset
                return
        if len(presets) >= MAX_PRESETS_PER_SERVICE:
            raise PresetError(
                f"Limite de {MAX_PRESETS_PER_SERVICE} presets para {preset.service_id}. "
                "Apague algum antes de gravar outro."
            )
        presets.append(preset)

    def remove(self, service_id: str, preset_id: str) -> bool:
        presets = self.by_service.get(service_id)
        if not presets:
            return False
        remaining = [p for p in presets if p.id != preset_id]
        if len(remaining) == len(presets):
            return False
        self.by_service[service_id] = remaining
        return True


def parse_store(data: dict[str, Any] | None) -> PresetStore:
    by_service: dict[str, list[Preset]] = {}
    for service_id, raw_presets in (data or {}).items():
        if not isinstance(raw_presets, list):
            raise PresetError(f"'{service_id}' deveria ser uma lista de presets")
        presets: list[Preset] = []
        for raw in raw_presets:
            if not isinstance(raw, dict) or "id" not in raw:
                raise PresetError(f"preset sem 'id' em {service_id}: {raw!r}")
            presets.append(
                Preset(
                    id=str(raw["id"]),
                    label=str(raw.get("label", raw["id"])),
                    service_id=str(service_id),
                    steps=tuple(parse_steps(raw.get("steps"))),
                    recorded_at=raw.get("recorded_at"),
                )
            )
        by_service[str(service_id)] = presets
    return PresetStore(by_service=by_service)


def load_presets(path: Path) -> PresetStore:
    """Arquivo ausente é normal: significa que você ainda não gravou nada."""
    if not path.exists():
        return PresetStore(by_service={})
    return parse_store(yaml.safe_load(path.read_text(encoding="utf-8")))


def save_presets(path: Path, store: PresetStore) -> None:
    payload = {
        service_id: [preset.as_yaml() for preset in presets]
        for service_id, presets in sorted(store.by_service.items())
        if presets
    }
    body = yaml.safe_dump(payload, allow_unicode=True, sort_keys=False, default_flow_style=False)
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(_HEADER + "\n" + body, encoding="utf-8")


def now_stamp() -> str:
    return datetime.now(UTC).strftime("%Y-%m-%d %H:%M")
